docs/fundamentals/code-analysis/rules/ca1000.md

```diff
@@ -0,0 +1,81 @@
+---
+title: "CA1000: Do not declare static members on generic types (code analysis)"
+description: "Learn about code analysis rule CA1000: Do not declare static members on generic types"
+ms.date: 03/11/2019
+ms.topic: reference
+f1_keywords:
+- CA1000
+- DoNotDeclareStaticMembersOnGenericTypes
+helpviewer_keywords:
+- DoNotDeclareStaticMembersOnGenericTypes
+- CA1000
+author: gewarren
+ms.author: gewarren
+dev_langs:
+- CSharp
+- VB
+---
+# CA1000: Do not declare static members on generic types
+
+|Item|Value|
+|-|-|
+|RuleId|CA1000|
+|Category|Microsoft.Design|
+|Fix is a breaking or non-breaking change|Breaking|
+
+## Cause
+
+A generic type contains a `static` (`Shared` in Visual Basic) member.
+
+By default, this rule only looks at externally visible types, but this is [configurable](#configurability).
+
+## Rule description
+
+When a `static` member of a generic type is called, the type argument must be specified for the type. When a generic instance member that does not support inference is called, the type argument must be specified for the member. The syntax for specifying the type argument in these two cases is different and easily confused, as the following calls demonstrate:
+
+```vb
+' Shared method in a generic type.
+GenericType(Of Integer).SharedMethod()
+
+' Generic instance method that does not support inference.
+someObject.GenericMethod(Of Integer)()
+```
+
+```csharp
+// Static method in a generic type.
+GenericType<int>.StaticMethod();
+
+// Generic instance method that does not support inference.
+someObject.GenericMethod<int>();
+```
+
+Generally, both of the prior declarations should be avoided so that the type argument does not have to be specified when the member is called. This results in a syntax for calling members in generics that is no different from the syntax for non-generics.
+
+## How to fix violations
+
+To fix a violation of this rule, remove the static member or change it to an instance member.
+
+## When to suppress warnings
+
+Do not suppress a warning from this rule. Providing generics in a syntax that is easy to understand and use reduces the time that is required to learn and increases the adoption rate of new libraries.
+
+## Configurability
+
+You can configure which parts of your codebase to run this rule on, based on their accessibility. For example, to specify that the rule should run only against the non-public API surface, add the following key-value pair to an *.editorconfig* file in your project:
+
+```ini
+dotnet_code_quality.ca1000.api_surface = private, internal
+```
+
+You can configure this option for just this rule, for all rules, or for all rules in this category (Design). For more information, see [Configure code analysis rules](../configure-rules.md).
```

Comment:
`../configure-rules.md` contains information about generic analyzer rule severity configuration, but we want this to link to option configuration page specific to CA rules, i.e. https://docs.microsoft.com/en-us/visualstudio/code-quality/configure-fxcop-analyzers?view=vs-2019. We should try to do one of the following:

1. Port https://docs.microsoft.com/en-us/visualstudio/code-quality/configure-fxcop-analyzers?view=vs-2019 to this repo and nest it under Configuration, so that the first overview sub-tab points to `configure-rules.md` and the second `Code quality` tab points to the ported page. Change the `[Configure code analysis rules]` hyperlink to point to the second page.
2. Change this hyperlink to point to https://docs.microsoft.com/en-us/visualstudio/code-quality/configure-fxcop-analyzers?view=vs-2019

I think (1) sounds more reasonable as there is no VS specific content at that page, but I'll let you decide if you would prefer the same, and if so, whether as part of this PR or a follow-up PR.

Reply:
Thanks Manish. What about the predefined rule sets that ship with the NuGet package. Are those in the .NET SDK too? Or should I leave that part out?

Reply:
The rulesets and editorconfigs do ship with `Microsoft.CodeAnalysis.NetAnalyzers` nuget package, and is specific to CA rules which ship in the SDK, so I think it makes sense to port it over too. You may want to remove the Windows specific nuget packages folder name and path as reported in https://github.com/dotnet/roslyn-analyzers/issues/4178#issuecomment-695762472